torchao/profiler/device_spec.py

```diff
@@ -0,0 +1,421 @@
+        torch.float64: 7e12,
+        torch.float32: 14e12,
+        torch.float16: 112e12,
+    },
+    "v100s pcie": {
+        torch.float64: 8.2e12,
+        torch.float32: 16.4e12,
+        torch.float16: 130e12,
+    },
+}
+
+
+# Adapted from https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/fabric/utilities/throughput.py
+def get_chip_name(device: int = 0) -> str:
+    device_name = torch.cuda.get_device_name(device)
+    chip = device_name.lower()
+
+    if "h100" in chip:
+        if "hbm3" in chip:
+            chip = "h100 sxm"
+        elif "nvl" in chip:
+            chip = "h100 nvl"
+        elif "pcie" in chip or "hbm2e" in chip:
+            chip = "h100 pcie"
+    elif "l4" in chip:
+        chip = "l40" if "tesla" in chip else "l4"
+    elif "geforce rtx" in chip:
+        number = chip.split(" ")[3]
+        extra = ""
+        if "super" in chip:
+            extra = " super"
+        elif "ti" in chip:
+            extra = " ti"
+        chip = f"rtx {number}{extra}"
+    elif "a6000" in chip:
+        chip = "a6000"
+    elif "a100" in chip:
+        chip = "a100"
+    elif "a40" in chip:
+        chip = "a40"
+    elif "a10g" in chip:
+        chip = "a10g"
+    elif "t4" in chip:
+        chip = "t4"
+    elif "quadro rtx 5000" in chip:
+        chip = "quadro rtx 5000"
+    elif "titan rtx" in chip:
+        chip = "titan rtx"
+    elif "v100-sxm" in chip:
+        chip = "v100 sxm"
+    elif "v100-pcie" in chip:
+        chip = "v100 pcie"
+    elif "v100s-pcie" in chip:
+        chip = "v100s pcie"
+    else:
+        chip = None
+    return chip
+
+
+def get_vram(device: int = 0) -> int:
+    device_props = torch.cuda.get_device_properties(device)
+    return device_props.total_memory
+
+
+def get_bandwidth(device: int = 0) -> int:
+    try:
+        from triton.testing import get_dram_gbps
+
+        bandwidth = get_dram_gbps(device) * 1e9
+    except ImportError:
+        print("Could not import triton to get DRAM Gbps. Please install triton")
+        bandwidth = None
+    return bandwidth
+
+
+def get_flops_by_dtype(chip_name: str) -> dict[torch.dtype, float]:
+    return _AVAILABLE_GPU_SPECS.get(chip_name, None)
+
+
+@dataclass
+class DeviceSpec:
+    """
+    Abstract device specs for theoretical peak performance calculations.
+
+    Fields will be auto-populated in __post_init__ if not already specified
+    and if data is available
+    - bandwidth (bytes /s)
+    - flops_per_s (FLOP / s)
+    - vram (bytes)
+    - dtype (torch.dtype) dtype used for theoretical peak performance
+    - flops_by_dtype (dict[Union[torch.dtype, str], float]): mapping from dtype to FLOP / s
+    """
+
+    device_type: str
+    name: Optional[str] = None
+    bandwidth: Optional[int] = None
+    flops_per_s: Optional[int] = None
+    vram: Optional[int] = None
+    dtype: Optional[torch.dtype] = None
+    flops_by_dtype: dict = field(default_factory=dict)
+
+    def _post_init_check(self):
```

Comment:
This example was failing on these asserts for me, one set of failures was because the code was based off an old branch and another set was because it seems like `device_spec` has non optional parameters like dtype and flops_by_dtype

```python
from torchao.profiler import CUDADeviceSpec, TransformerPerformanceCounter
import torch

# Device spec is a dataclass that contains device info such as name, peak bandwidth, peak FLOPs
# If these fields are not manually specified, they will be automatically populated using
# CUDA runtime functions exposed by `torch.cuda` and `triton.runtime.driver`  
device_spec = CUDADeviceSpec()

# The manager object tracks latency, data movement (bytes), and FLOPs across multiple contexts and
# maintains performance metrics for each context and in aggregate.
manager = TransformerPerformanceCounter(device_spec=device_spec)

# Prefill
with manager.count(label="prefill", num_tokens=x.numel()):
    out = model(encoded_prompt)
# Print recorded stats for "prefill" context
manager.print_summary(labels=["prefill"], show=True) 

# Decode
with manager.count(label="decode", num_tokens=1):
    out = model(out[-1])
# Print recorded stats for "decode" context
manager.print_summary(labels=["decode"], show=True) 

# Print accumulated stats across all contexts
manager.print_summary(show=True) # 

```